# **San Francisco Crime Project**

- **Author:** Muhammad Jawad [@mjawad17]()
- **Description:** Data analysis, exploration, visualization, and data mining on crime in SF
- **Original dataset:** [SF Gov Crime dataset](https://data.sfgov.org/Public-Safety/Police-Department-Incident-Reports-Historical-2003/tmnf-yvry/about_data)
- **Kaggle dataset:** [Kaggle SF Crime](https://www.kaggle.com/competitions/sf-crime/overview)

<div style="border: 2px solid #4CAF50; border-radius: 10px; padding: 20px; background-color: #f9f9f9; font-family: Arial, sans-serif; color: #333;">
    <h2 style="color: #4CAF50;">Author Overview</h2>
    <!-- Round image centered -->
    <img src="https://avatars.githubusercontent.com/u/77524488?s=400&u=5ee60100c5daf1eb876be2bc80aaa0e9e85969c3&v=4" alt="Author Image" style="border-radius: 12%; width: 200px; height: 200px; margin-bottom: 20px; border: 2px solid green; display: block; margin-left: auto; margin-right: auto;">
    <p>I am <strong>Muhammad Jawad</strong>, a passionate data analyst dedicated to leveraging data to drive meaningful insights and support decision-making. With a strong foundation in computer science, I continuously seek to enhance my skills in analytical thinking and data-driven solutions. 📊</p>
    <h3 style="color: #4CAF50;">What Do I Know?</h3>
    <p>I excel at extracting valuable insights from data using Python, SQL, and data visualization tools like Power BI and Tableau. My experience includes analyzing insurance and medical data, as well as performing exploratory data analysis to uncover trends and patterns. I'm committed to improving my statistical analysis and reporting abilities to solve complex problems effectively. 📚</p>
    <h3 style="color: #4CAF50;">What Am I Doing Right Now?</h3>
    <p>Currently, I am focused on expanding my knowledge in data science, particularly in machine learning and advanced analytics. I am eager to apply my theoretical knowledge to real-world projects that challenge me and contribute to organizational success. 🎯</p>
    <h3 style="color: #4CAF50;">My Goal:</h3>
    <p>To utilize my data analysis and data science skills to create value for companies by supporting their growth objectives. I am enthusiastic about learning new concepts and sharing my experiences with others. If you’re interested in discussing projects, collaborating, or exchanging ideas, I would love to connect!</p>
</div>

# Table of Contents
- Introduction
    - SF Crime Dataset
- Basic Preparation
    - Import libraries
    - Load data
- Data Exploration/Analysis Extension
- Data Preprocessing
    - Data Imputation/Removal
    - Feature Engineering
    - Feature Encoding
- Build Machine Learning Models
    - Train different baseline models
    - Analyze results
- Model Selection
- Hyperparameter tuning
- Train Model with optimal hyperparameters
- Feature Selection
    - Feature Importance
    - Feature Removal
- Train Final Model
- Model Evaluation
- Summary
<!-- - Kaggle Submission -->
- Conclusion

# Introduction

## SF Crime Dataset

This dataset includes information about crime incidents reported by the **San Francisco Police Department (SFPD)**. It covers data from _January 1, 2003, to May 13, 2015_.

The dataset is divided into **two groups**: a training set and a test set. These sets rotate weekly. This means that in odd weeks (like the 1st, 3rd, 5th, and 7th weeks), the data is used for the test set. In even weeks (like the 2nd, 4th, 6th, and 8th weeks), the data is used for the training set.

The main **objective** of this dataset is to predict the category of crime that took place in San Francisco based on the available information.

### Data Fields
- **Dates** - timestamp of the crime incident
- **Category** - category of the crime incident (only in train.csv). This is the target variable you are going to predict.
- **Descript** - detailed description of the crime incident (only in train.csv)
- **DayOfWeek** - the day of the week
- **PdDistrict** - name of the Police Department District
- **Resolution** - how the crime incident was resolved (only in train.csv)
- **Address** - the approximate street address of the crime incident
- **X** - Longitude
- **Y** - Latitude

---

In this Jupyter notebook, I will take you through the entire process of creating a machine learning model using the open-source San Francisco Crime dataset. This will be a step-by-step journey that includes several important stages.

First, I will explore and analyze the data to understand its structure and contents. This is a crucial step that helps identify patterns and insights within the data. Next, I will preprocess the data, which is a significant part of this project. This step involves cleaning the data and performing feature engineering to create useful variables for the model.

After preparing the data, I will try out different machine learning algorithms to see which one works best for this dataset. I will determine the most effective model and then fine-tune its hyperparameters to improve its performance. Finally, I will evaluate the chosen model using a metric called multiclass log loss to assess how well it predicts the categories of crime.

Since this project is based on an older Kaggle competition, I will avoid looking for external resources or past Kaggle notebooks. My goal is to enhance my coding skills for an end-to-end data science project and to become more familiar with Python data science libraries. I also hope to uncover interesting insights and discover cool patterns while working with this dataset. So, let’s get started!

# Basic Preparation

## Import Libraries

In [1]:
pip install scikit-optimize

Note: you may need to restart the kernel to use updated packages.


In [2]:
__author__ = "Muhammad Jawad (https://github.com/mj-awad17)"

# linear algebra
import numpy as np

# data manipulation
import pandas as pd

# plotting
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib import style
%matplotlib inline
style.use('ggplot')

# preprocessing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder, OneHotEncoder

# machince learning algorithms
from sklearn import linear_model
from sklearn.linear_model import LogisticRegression, SGDClassifier, Perceptron
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, AdaBoostClassifier
from sklearn.svm import SVC, LinearSVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB
from sklearn.neural_network import MLPClassifier
from xgboost import XGBClassifier
import xgboost as xgb

# model evaluation metrics
from sklearn.metrics import log_loss
from sklearn.model_selection import cross_val_score

# model selection and tuning
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV, StratifiedGroupKFold
from skopt import BayesSearchCV
from skopt.space import Real, Categorical, Integer

# clustering
from sklearn.cluster import KMeans

# mathematical fundtions
import math

# ignore warnings
import warnings
warnings.filterwarnings("ignore")

# display all columns in pandas dataframe
pd.set_option('display.max_columns', None)

## Load Data

In [3]:
train = pd.read_csv('../../All/train.csv')
test = pd.read_csv('../../All/test.csv')

In [4]:
# info of train data
print("Train Data Info:")
print(train.info())

Train Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 9 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   Dates       878049 non-null  object 
 1   Category    878049 non-null  object 
 2   Descript    878049 non-null  object 
 3   DayOfWeek   878049 non-null  object 
 4   PdDistrict  878049 non-null  object 
 5   Resolution  878049 non-null  object 
 6   Address     878049 non-null  object 
 7   X           878049 non-null  float64
 8   Y           878049 non-null  float64
dtypes: float64(2), object(7)
memory usage: 60.3+ MB
None


# Data Exploration & Analysis Extension

- Complete data exploration & visualizations are done in jupyter notebook: [sf-crime-data-exploration.ipynb](sf-crime-exploration.ipynb)
- This dataset suffers from **imbalanced** classes (TREA has 6 occurrences while LARCENY/THEFT has 1,749,000 occurrences)
    - There are a couple ways to deal with **imbalanced** classes, such as:
        - Changing performance metric (Do not use accuracy, use a confusion matrix, precision, recall, F1 score, ROC curves)
        - Resample dataset (Oversample under-represented classes, and undersample over-represented classes)
        - Try different ML algorithms that can handle imbalanced classes
            - Decision Trees (Random Forests/XGBoost) often perform well on imbalanced classes (due to splitting rules)

In [5]:
train.head()

,Dates,Category,Descript,DayOfWeek,PdDistrict,Resolution,Address,X,Y
0,2015-05-13 23:53:00,WARRANTS,WARRANT ARREST,Wednesday,NORTHERN,"ARREST, BOOKED",OAK ST / LAGUNA ST,-122.425892,37.774599
1,2015-05-13 23:53:00,OTHER OFFENSES,TRAFFIC VIOLATION ARREST,Wednesday,NORTHERN,"ARREST, BOOKED",OAK ST / LAGUNA ST,-122.425892,37.774599
2,2015-05-13 23:33:00,OTHER OFFENSES,TRAFFIC VIOLATION ARREST,Wednesday,NORTHERN,"ARREST, BOOKED",VANNESS AV / GREENWICH ST,-122.424363,37.800414
3,2015-05-13 23:30:00,LARCENY/THEFT,GRAND THEFT FROM LOCKED AUTO,Wednesday,NORTHERN,NONE,1500 Block of LOMBARD ST,-122.426995,37.800873
4,2015-05-13 23:30:00,LARCENY/THEFT,GRAND THEFT FROM LOCKED AUTO,Wednesday,PARK,NONE,100 Block of BRODERICK ST,-122.438738,37.771541


In [6]:
train.columns.values

array(['Dates', 'Category', 'Descript', 'DayOfWeek', 'PdDistrict',
       'Resolution', 'Address', 'X', 'Y'], dtype=object)

### Things we Observed so far:
- 878,049 instances in training set (or recorded crime instances in SF)
- 9 columns (8 potential features + 1 label (Category))

**Data types:**

- 2 columns with float values
- 7 objects

There are no null (NaN) values (Yaha!)

In [7]:
# Count number of observations for each crime 
train['Category'].value_counts()

Category
LARCENY/THEFT                  174900
OTHER OFFENSES                 126182
NON-CRIMINAL                    92304
ASSAULT                         76876
DRUG/NARCOTIC                   53971
VEHICLE THEFT                   53781
VANDALISM                       44725
WARRANTS                        42214
BURGLARY                        36755
SUSPICIOUS OCC                  31414
MISSING PERSON                  25989
ROBBERY                         23000
FRAUD                           16679
FORGERY/COUNTERFEITING          10609
SECONDARY CODES                  9985
WEAPON LAWS                      8555
PROSTITUTION                     7484
TRESPASS                         7326
STOLEN PROPERTY                  4540
SEX OFFENSES FORCIBLE            4388
DISORDERLY CONDUCT               4320
DRUNKENNESS                      4280
RECOVERED VEHICLE                3138
KIDNAPPING                       2341
DRIVING UNDER THE INFLUENCE      2268
RUNAWAY                          1946
LIQ

In [8]:
# Count number of observations for each PdDistrict
train['PdDistrict'].value_counts()

PdDistrict
SOUTHERN      157182
MISSION       119908
NORTHERN      105296
BAYVIEW        89431
CENTRAL        85460
TENDERLOIN     81809
INGLESIDE      78845
TARAVAL        65596
PARK           49313
RICHMOND       45209
Name: count, dtype: int64

In [9]:
# Count number of observations for each DayOfWeek
train['DayOfWeek'].value_counts()

DayOfWeek
Friday       133734
Wednesday    129211
Saturday     126810
Thursday     125038
Tuesday      124965
Monday       121584
Sunday       116707
Name: count, dtype: int64

In [10]:
# Count number of observations for each DayOfWeek
train['DayOfWeek'].value_counts()

DayOfWeek
Friday       133734
Wednesday    129211
Saturday     126810
Thursday     125038
Tuesday      124965
Monday       121584
Sunday       116707
Name: count, dtype: int64

In [11]:
# Count number of observations for each Resolution
train['Resolution'].value_counts()

Resolution
NONE                                      526790
ARREST, BOOKED                            206403
ARREST, CITED                              77004
LOCATED                                    17101
PSYCHOPATHIC CASE                          14534
UNFOUNDED                                   9585
JUVENILE BOOKED                             5564
COMPLAINANT REFUSES TO PROSECUTE            3976
DISTRICT ATTORNEY REFUSES TO PROSECUTE      3934
NOT PROSECUTED                              3714
JUVENILE CITED                              3332
PROSECUTED BY OUTSIDE AGENCY                2504
EXCEPTIONAL CLEARANCE                       1530
JUVENILE ADMONISHED                         1455
JUVENILE DIVERTED                            355
CLEARED-CONTACT JUVENILE FOR MORE INFO       217
PROSECUTED FOR LESSER OFFENSE                 51
Name: count, dtype: int64

In [12]:
# describe the train data
train.describe().T

,count,mean,std,min,25%,50%,75%,max
X,878049.0,-122.422616,0.030354,-122.513642,-122.432952,-122.416420,-122.406959,-120.5
Y,878049.0,37.771020,0.456893,37.707879,37.752427,37.775421,37.784369,90.0


> There seems to be an invalid coordinates (max) 90 (latitude) or -120.5 (longitude) does not seem to be a valid coordinate in San Francisco. We must fix these values for this feature.

# Data Preprocessing

- Data cleaning
    - imputation or removal of outlier values
- Feature Engineering (Feature Creation)
- Feature Encoding
    - Integer encode or label encode ordinal categorical features that maintain order (Year, Business Quarter, Block/Street Number)
    - Usually:
        - One hot encode nominal categorical features (DayOfWeek, PdDistrict, StreetType, Category) 
            - mainly for logistic regression
        - However, Random Forests & Boosting algorithms can handle nominal categorical features directly, so we just integer encode these features.

## Data Cleaning

In [13]:
train[train['Y'] == train['Y'].max()]

,Dates,Category,Descript,DayOfWeek,PdDistrict,Resolution,Address,X,Y
660485,2005-12-30 17:00:00,LARCENY/THEFT,GRAND THEFT FROM LOCKED AUTO,Friday,TENDERLOIN,NONE,5THSTNORTH ST / OFARRELL ST,-120.5,90.0
660711,2005-12-30 00:34:00,ASSAULT,INFLICT INJURY ON COHABITEE,Friday,BAYVIEW,"ARREST, BOOKED",JAMESLICKFREEWAY HY / SILVER AV,-120.5,90.0
660712,2005-12-30 00:34:00,ASSAULT,AGGRAVATED ASSAULT WITH BODILY FORCE,Friday,BAYVIEW,"ARREST, BOOKED",JAMESLICKFREEWAY HY / SILVER AV,-120.5,90.0
661106,2005-12-29 00:07:00,NON-CRIMINAL,"AIDED CASE, MENTAL DISTURBED",Thursday,TENDERLOIN,PSYCHOPATHIC CASE,5THSTNORTH ST / EDDY ST,-120.5,90.0
666430,2005-11-30 11:25:00,OTHER OFFENSES,TRAFFIC VIOLATION,Wednesday,TENDERLOIN,"ARREST, CITED",5THSTNORTH ST / ELLIS ST,-120.5,90.0
...,...,...,...,...,...,...,...,...,...
844995,2003-06-11 08:49:00,OTHER OFFENSES,"DRIVERS LICENSE, SUSPENDED OR REVOKED",Wednesday,INGLESIDE,"ARREST, CITED",JAMES LICK FREEWAY HY / CESAR CHAVEZ ST,-120.5,90.0
845842,2003-06-09 09:25:00,OTHER OFFENSES,"DRIVERS LICENSE, SUSPENDED OR REVOKED",Monday,INGLESIDE,"ARREST, CITED",JAMES LICK FREEWAY HY / CESAR CHAVEZ ST,-120.5,90.0
852880,2003-05-02 01:00:00,SEX OFFENSES FORCIBLE,"FORCIBLE RAPE, BODILY FORCE",Friday,SOUTHERN,COMPLAINANT REFUSES TO PROSECUTE,3RD ST / JAMES LICK FREEWAY HY,-120.5,90.0
857248,2003-04-14 16:30:00,ROBBERY,"ROBBERY ON THE STREET, STRONGARM",Monday,BAYVIEW,COMPLAINANT REFUSES TO PROSECUTE,GILMAN AV / FITCH ST,-120.5,90.0


I notice that there are `108` **rows** with incorrect coordinates, and they seem to be the exact same two coordinates (90, -120.5). There are many ways to handle this. We need to do data **imputation**, which can be done several ways. For now, I will randomly sample from a normal distribution with the range of a standard deviation from the mean. However, I could use a linear regression model to predict the latitude and longitude values (based on other variables such as PD district?) and use that to impute the bad / inconsistent data points.

Another method is to completely remove this data. Since I already have a lot of data, and I do not want this incorrect data to affect my results, I could remove them. However, I will stick with data imputation.

In [14]:
# replace the max value of Y and X with NaN: both Train and test data
train['Y'].replace(to_replace= train['Y'].max() ,value=np.nan, inplace=True)
train['X'].replace(to_replace= train['X'].max() ,value=np.nan, inplace=True)
test['Y'].replace(to_replace= test['Y'].max() ,value=np.nan, inplace=True)
test['X'].replace(to_replace= test['X'].max() ,value=np.nan, inplace=True)

In [15]:
print("Train Dataset:", train.isnull().sum())
print("\nTest Dataset:", test.isnull().sum())

Train Dataset: Dates          0
Category       0
Descript       0
DayOfWeek      0
PdDistrict     0
Resolution     0
Address        0
X             67
Y             67
dtype: int64

Test Dataset: Id             0
Dates          0
DayOfWeek      0
PdDistrict     0
Address        0
X             76
Y             76
dtype: int64


In [16]:
# combine train and test data for preprocessing
df = [train, test]

# iterate through the dataframes for statistical methods used
for data in df:
    mean_X = data['X'].mean()
    std_X = data['X'].std()
    mean_Y = data['Y'].mean()
    std_Y = data['Y'].std()
    max_x = mean_X + std_X
    min_x = mean_X - std_X
    max_y = mean_Y + std_Y
    min_y = mean_Y - std_Y

    # both X and Y are same so just using Y
    is_null = data['Y'].isnull().sum()

    random_X = (max_x - min_x) * np.random.randn(is_null) + min_x
    random_Y = (max_y - min_y) * np.random.randn(is_null) + min_y

    X_slice = data['X'].copy()
    Y_slice = data['Y'].copy()
    X_slice[np.isnan(X_slice)] = random_X
    Y_slice[np.isnan(Y_slice)] = random_Y
    data['X'] = X_slice
    data['Y'] = Y_slice

In [17]:
train[['X', 'Y']].describe().T

,count,mean,std,min,25%,50%,75%,max
X,878049.0,-122.422765,0.025288,-122.556583,-122.432952,-122.416422,-122.406959,-122.333414
Y,878049.0,37.767033,0.024167,37.659588,37.752427,37.775421,37.784368,37.852662


In [18]:
len(test)

884262

## Feature Engineering
- Let's create some new features from the data that exists in the current feature space
- There are a couple categories of features:
    - Temporal features
    - Spatial features

### Temporal Features
We want to have a column for Time, so we must parse through the 'Dates' feature to create the 'Time' feature

In [19]:
# Transform the Date into a python datetime object.
train["Dates"] = pd.to_datetime(train["Dates"], format="%Y-%m-%d %H:%M:%S")
test["Dates"] = pd.to_datetime(test["Dates"], format="%Y-%m-%d %H:%M:%S")

In [20]:
# minutes
train["Minutes"] = train["Dates"].map(lambda x: x.minute)
test["Minutes"] = test["Dates"].map(lambda x: x.minute)
# hours
train["Hours"] = train["Dates"].map(lambda x: x.hour)
test["Hours"] = test["Dates"].map(lambda x: x.hour)
# days
train["Days"] = train["Dates"].map(lambda x: x.day)
test["Days"] = test["Dates"].map(lambda x: x.day)
# months
train["Months"] = train["Dates"].map(lambda x: x.month)
test["Months"] = test["Dates"].map(lambda x: x.month)
# years
train["Years"] = train["Dates"].map(lambda x: x.year)
test["Years"] = test["Dates"].map(lambda x: x.year)

In [21]:
# Hour Zone 0 - Pass midnight, 1 - morning, 2 - afternoon, 3 - dinner / sun set, 4 - night
def hour_zone(hour):
    if hour < 1:
        return 0
    elif hour < 6:
        return 1
    elif hour < 12:
        return 2
    elif hour < 18:
        return 3
    else:
        return 4

train["HourZone"] = train["Hours"].map(hour_zone)
test["HourZone"] = test["Hours"].map(hour_zone)

In [22]:
# Add Week of Year
train["WeekOfYear"] = train["Dates"].map(lambda x: int(x.weekofyear / 2) - 1)
test["WeekOfYear"] = test["Dates"].map(lambda x: int(x.weekofyear / 2))

print(sorted(train['WeekOfYear'].unique()))
print(sorted(test['WeekOfYear'].unique()))

[np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12), np.int64(13), np.int64(14), np.int64(15), np.int64(16), np.int64(17), np.int64(18), np.int64(19), np.int64(20), np.int64(21), np.int64(22), np.int64(23), np.int64(24), np.int64(25)]
[np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12), np.int64(13), np.int64(14), np.int64(15), np.int64(16), np.int64(17), np.int64(18), np.int64(19), np.int64(20), np.int64(21), np.int64(22), np.int64(23), np.int64(24), np.int64(25), np.int64(26)]


In [23]:
train.head()

,Dates,Category,Descript,DayOfWeek,PdDistrict,Resolution,Address,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear
0,2015-05-13 23:53:00,WARRANTS,WARRANT ARREST,Wednesday,NORTHERN,"ARREST, BOOKED",OAK ST / LAGUNA ST,-122.425892,37.774599,53,23,13,5,2015,4,9
1,2015-05-13 23:53:00,OTHER OFFENSES,TRAFFIC VIOLATION ARREST,Wednesday,NORTHERN,"ARREST, BOOKED",OAK ST / LAGUNA ST,-122.425892,37.774599,53,23,13,5,2015,4,9
2,2015-05-13 23:33:00,OTHER OFFENSES,TRAFFIC VIOLATION ARREST,Wednesday,NORTHERN,"ARREST, BOOKED",VANNESS AV / GREENWICH ST,-122.424363,37.800414,33,23,13,5,2015,4,9
3,2015-05-13 23:30:00,LARCENY/THEFT,GRAND THEFT FROM LOCKED AUTO,Wednesday,NORTHERN,NONE,1500 Block of LOMBARD ST,-122.426995,37.800873,30,23,13,5,2015,4,9
4,2015-05-13 23:30:00,LARCENY/THEFT,GRAND THEFT FROM LOCKED AUTO,Wednesday,PARK,NONE,100 Block of BRODERICK ST,-122.438738,37.771541,30,23,13,5,2015,4,9


### Holiday Feature
- Certain crimes may be more apparent on holidays

In [24]:
from pandas.tseries.holiday import USFederalHolidayCalendar as calendar

# Training set
cal = calendar()
holidays = cal.holidays(start=train['Dates'].min(), end=train['Dates'].max())
train['Holiday'] = train['Dates'].dt.date.astype('datetime64[ns]').isin(holidays)
# Test set
cal = calendar()
holidays = cal.holidays(start=test['Dates'].min(), end=test['Dates'].max())
test['Holiday'] = test['Dates'].dt.date.astype('datetime64[ns]').isin(holidays)

In [25]:
print("Train:", len(train[train['Holiday'] == True]))
# print("Test:", len(test[test['Holiday'] == True]))

Train: 25653


### Business Hours Feature
- There should be an effect of business hours on the type of crime committed
- Let's create a binary feature where:
    - 1 is typical business hours [8:00AM - 6:00PM]
    - 0 is not business hours [6:01PM - 7:59 AM]

In [26]:
from datetime import datetime, time

def time_in_range(start, end, x):
    """Return true if x is in the inclusive range [start, end]"""
    if start <= end:
        return start <= x <= end
    else:
        return start <= x or x <= end

def map_business_hours(date):
    
    # Convert military time to AM & PM
    time_parsed = date.time()
    business_start = time(8, 0, 0)
    business_end = time(18, 0, 0)
    
    if time_in_range(business_start, business_end, time_parsed):
        return 1
    else:
        return 0
    
train['BusinessHour'] = train['Dates'].map(map_business_hours).astype('uint8')
test['BusinessHour'] = test['Dates'].map(map_business_hours).astype('uint8')

In [27]:
train['BusinessHour'].value_counts()

BusinessHour
1    455215
0    422834
Name: count, dtype: int64

In [28]:
train.sample(5)

,Dates,Category,Descript,DayOfWeek,PdDistrict,Resolution,Address,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour
21214,2015-02-03 05:30:00,VEHICLE THEFT,STOLEN AUTOMOBILE,Tuesday,TARAVAL,NONE,JUNIPERO SERRA BL / HOLLOWAY AV,-122.472322,37.721622,30,5,3,2,2015,1,2,False,0
446293,2009-02-02 15:55:00,FORGERY/COUNTERFEITING,"COUNTERFEITING, COINS OR NOTES",Monday,SOUTHERN,"ARREST, BOOKED",1200 Block of MARKET ST,-122.415444,37.778025,55,15,2,2,2009,3,2,False,1
370175,2010-03-14 19:45:00,BURGLARY,"BURGLARY OF STORE, UNLAWFUL ENTRY",Sunday,MISSION,"ARREST, BOOKED",2300 Block of 16TH ST,-122.409530,37.765718,45,19,14,3,2010,4,4,False,0
423233,2009-05-28 09:30:00,VANDALISM,"MALICIOUS MISCHIEF, VANDALISM OF VEHICLES",Thursday,NORTHERN,NONE,LARKIN ST / WASHINGTON ST,-122.419583,37.793466,30,9,28,5,2009,2,10,False,1
780880,2004-04-18 01:00:00,LARCENY/THEFT,GRAND THEFT OF PROPERTY,Sunday,CENTRAL,NONE,400 Block of BROADWAY ST,-122.404282,37.798225,0,1,18,4,2004,1,7,False,0


### Business Quarter (Removed)
- Business Quarter might have an effect on what types of crimes are commited
- Q1 = 1 (Jan. - March) Q2 = 2 (April - June), Q3 = 3 (July - Sept.), Q4 = 4 (Oct. - Dec.)

In [29]:
train.columns

Index(['Dates', 'Category', 'Descript', 'DayOfWeek', 'PdDistrict',
       'Resolution', 'Address', 'X', 'Y', 'Minutes', 'Hours', 'Days', 'Months',
       'Years', 'HourZone', 'WeekOfYear', 'Holiday', 'BusinessHour'],
      dtype='object')

In [30]:
def map_business_quarter(month):
    
    if month in [1, 2, 3]:
        return 1
    elif month in [4, 5, 6]:
        return 2
    elif month in [7, 8, 9]:
        return 3
    elif month in [10, 11, 12]:
        return 4
    
train['Quarter'] = train['Months'].map(map_business_quarter)
test['Quarter'] = test['Months'].map(map_business_quarter)

### Season
The season feature may affect what type of crimes are commited.

- 1 = Winter, 2 = Spring, 3 = Summer, 4 = Fall

In [31]:
train['Season']=(train['Months']%12 + 3)//3
test['Season']=(test['Months']%12 + 3)//3

In [32]:
train.sample(5)

,Dates,Category,Descript,DayOfWeek,PdDistrict,Resolution,Address,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour,Quarter,Season
676977,2005-10-07 11:24:00,WARRANTS,ENROUTE TO DEPARTMENT OF CORRECTIONS,Friday,TENDERLOIN,"ARREST, BOOKED",200 Block of TURK ST,-122.413238,37.782843,24,11,7,10,2005,2,19,False,1,4,4
335577,2010-09-22 10:35:00,TRESPASS,TRESPASSING,Wednesday,SOUTHERN,"ARREST, CITED",1400 Block of MARKET ST,-122.418481,37.775898,35,10,22,9,2010,2,18,False,1,3,4
581370,2007-02-21 12:51:00,DRUG/NARCOTIC,POSSESSION OF BASE/ROCK COCAINE FOR SALE,Wednesday,BAYVIEW,"ARREST, BOOKED",3RD ST / OAKDALE AV,-122.390698,37.734797,51,12,21,2,2007,3,3,False,1,1,1
88712,2014-03-08 04:00:00,LARCENY/THEFT,GRAND THEFT FROM UNLOCKED AUTO,Saturday,MISSION,NONE,VALENCIA ST / DUBOCE AV,-122.422367,37.769868,0,4,8,3,2014,1,4,False,0,1,2
846048,2003-06-01 18:10:00,OTHER OFFENSES,MISCELLANEOUS INVESTIGATION,Sunday,SOUTHERN,NONE,600 Block of 2ND ST,-122.391238,37.781253,10,18,1,6,2003,4,10,False,0,2,3


### Weekend
Weekends may have effect on what types of crimes are commmited
- Weekday = 0, Weekend =1

In [33]:
# Weekend Feature
# Weekday = 0, Weekend = 1
days = {'Monday':0 ,'Tuesday':0 ,'Wednesday':0 ,'Thursday':0 ,'Friday':0, 'Saturday':1 ,'Sunday':1}

train['Weekend'] = train['DayOfWeek'].replace(days).astype('uint8')
test['Weekend'] = test['DayOfWeek'].replace(days).astype('uint8')

## Spatial Features
### Street Type
The street type can have an effect on what type of crime is committed, so we want to extract the street type from the 'Address' feature.

We have avenues, streets, ways, boulevards, highways, courts, walks, plazas, and differet number of intersections of roads/streets (Addresses with /).

In [34]:
train['Address'].value_counts().index

Index(['800 Block of BRYANT ST', '800 Block of MARKET ST',
       '2000 Block of MISSION ST', '1000 Block of POTRERO AV',
       '900 Block of MARKET ST', '0 Block of TURK ST', '0 Block of 6TH ST',
       '300 Block of ELLIS ST', '400 Block of ELLIS ST',
       '16TH ST / MISSION ST',
       ...
       'KENWOOD WY / UPLAND DR', 'AVALON AV / PERU AV',
       'OFARRELL ST / ANZA VISTA AV', '400 Block of POINT LOBOS AV',
       'KNOLLVIEW WY / CITYVIEW WY', 'VAN DYKE AV / LANE ST',
       'VAN NESS AV / IVY ST', 'STANYAN ST / BELGRAVE AV',
       '0 Block of NORIEGA ST', '200 Block of JESSIE ST'],
      dtype='object', name='Address', length=23228)

In [35]:
import re

    
def find_streets(address):
    street_types = ['AV', 'ST', 'CT', 'PZ', 'LN', 'DR', 'PL', 'HY', 
                    'FY', 'WY', 'TR', 'RD', 'BL', 'WAY', 'CR', 'AL', 'I-80',  
                    'RW', 'WK','EL CAMINO DEL MAR']
    street_pattern = '|'.join(street_types)
    streets = re.findall(street_pattern, address)
    if len(streets) == 0:
        # Debug
#         print(address)
        return 'OTHER'
    elif len(streets) == 1:
        return streets[0]
    else:
#         print(address)
        return 'INT'

train['StreetType'] = train['Address'].map(find_streets)
test['StreetType'] = test['Address'].map(find_streets)

In [36]:
train['StreetType'].value_counts()

StreetType
INT                  389995
ST                   358797
AV                    92467
BL                    13074
DR                     8200
WY                     4063
RD                     2384
PZ                     2347
CT                     2059
LN                     1356
PL                      863
HY                      819
TR                      766
I-80                    322
CR                      291
AL                      150
WAY                      55
EL CAMINO DEL MAR        21
OTHER                    14
WK                        5
RW                        1
Name: count, dtype: int64

In [37]:
train['StreetType'].isnull().sum()

np.int64(0)

In [38]:
train.head(8)

,Dates,Category,Descript,DayOfWeek,PdDistrict,Resolution,Address,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour,Quarter,Season,Weekend,StreetType
0,2015-05-13 23:53:00,WARRANTS,WARRANT ARREST,Wednesday,NORTHERN,"ARREST, BOOKED",OAK ST / LAGUNA ST,-122.425892,37.774599,53,23,13,5,2015,4,9,False,0,2,2,0,INT
1,2015-05-13 23:53:00,OTHER OFFENSES,TRAFFIC VIOLATION ARREST,Wednesday,NORTHERN,"ARREST, BOOKED",OAK ST / LAGUNA ST,-122.425892,37.774599,53,23,13,5,2015,4,9,False,0,2,2,0,INT
2,2015-05-13 23:33:00,OTHER OFFENSES,TRAFFIC VIOLATION ARREST,Wednesday,NORTHERN,"ARREST, BOOKED",VANNESS AV / GREENWICH ST,-122.424363,37.800414,33,23,13,5,2015,4,9,False,0,2,2,0,INT
3,2015-05-13 23:30:00,LARCENY/THEFT,GRAND THEFT FROM LOCKED AUTO,Wednesday,NORTHERN,NONE,1500 Block of LOMBARD ST,-122.426995,37.800873,30,23,13,5,2015,4,9,False,0,2,2,0,INT
4,2015-05-13 23:30:00,LARCENY/THEFT,GRAND THEFT FROM LOCKED AUTO,Wednesday,PARK,NONE,100 Block of BRODERICK ST,-122.438738,37.771541,30,23,13,5,2015,4,9,False,0,2,2,0,ST
5,2015-05-13 23:30:00,LARCENY/THEFT,GRAND THEFT FROM UNLOCKED AUTO,Wednesday,INGLESIDE,NONE,0 Block of TEDDY AV,-122.403252,37.713431,30,23,13,5,2015,4,9,False,0,2,2,0,AV
6,2015-05-13 23:30:00,VEHICLE THEFT,STOLEN AUTOMOBILE,Wednesday,INGLESIDE,NONE,AVALON AV / PERU AV,-122.423327,37.725138,30,23,13,5,2015,4,9,False,0,2,2,0,INT
7,2015-05-13 23:30:00,VEHICLE THEFT,STOLEN AUTOMOBILE,Wednesday,BAYVIEW,NONE,KIRKWOOD AV / DONAHUE ST,-122.371274,37.727564,30,23,13,5,2015,4,9,False,0,2,2,0,INT


### Block Features (Removed)
- Let's explore and create the block feature, since we saw it a lot in the address features
- Binary feature
    - Categorize address that contains 'Block', as having a block, and if no block exists, we will assign to 0.
- 617231 addresses with blocks
- 260818 addresses with no blocks

In [39]:
def find_block(address):
    block_pattern = 'Block'
    blocks = re.search(block_pattern, address)
    if blocks:
#         print(address)
        return 1
    else:
#         print(address)
        return 0

In [40]:
train['Block'] = train['Address'].map(find_block)
test['Block'] = test['Address'].map(find_block)
train['Block'].value_counts()

Block
1    617231
0    260818
Name: count, dtype: int64

### Block Number Feature
- Let's explore the block number from address
- Block number has ordinal data type (order matters), and has spatial significance
- It seems all the block numbers are in intervals of 100
- How to categorize
    - Addresses that do not have a block number will be categorized as 0
    - Addresses with block number will be divided by 100, and added by 1 for mapping (0 is saved for addresses with no block number)
- 85 unique block numbers (including 1 where there is no block number)

In [41]:
def find_block_number(address):
    block_num_pattern = '[0-9]+\s[Block]'
    block_num = re.search(block_num_pattern, address)
    if block_num:
        # print(address)
        num_pattern = '[0-9]+'
        block_no_pos = re.search(num_pattern, address)
        # Get integer of found regular expression
        block_no = int(block_no_pos.group())
        # Convert block number by dividing by 100 and adding 1 (0 = addresses with no block)
        block_map = (block_no // 100) + 1
        # print(block_map)
        return block_map
    else:
        # print(address)
        return 0


train['BlockNo'] = train['Address'].map(find_block_number)
test['BlockNo'] = test['Address'].map(find_block_number)

In [42]:
train['BlockNo'].value_counts()

BlockNo
0     260818
1      76325
2      51917
9      51718
3      38407
       ...  
82         7
79         5
81         4
84         4
80         3
Name: count, Length: 85, dtype: int64

## X, Y Coordinates
- Normalize and scale the X and Y coordinates
- I use K-Means clustering to create a new feature for the longitude and latitude by grouping clusters of points based on Euclidean distances.
- X = longitude, Y = latitude
- I also extract more spatial features from the X, Y coordinates by transforming them from the cartesian space to the polar space (Reference)
    1. three variants of rotated Cartesian coordinates (rotated by 30, 45, 60 degree each)
    2. Polar coordinates (i.e. the 'r' and the angle 'theta')
    3. The approach makes some intuitive sense i.e. that having such features should help in extracting some more spatial information (than relying on the current x-y alone)

In [43]:
# Normalize X and Y
print('There are %d unique longitude values, %d unique latitude values' % (train['X'].nunique(), 
                                                                           train['Y'].nunique()))

xy_scaler = StandardScaler().fit(train[['X', 'Y']])
train[['X', 'Y']] = xy_scaler.transform(train[['X', 'Y']])
test[['X', 'Y']] = xy_scaler.transform(test[['X', 'Y']])

There are 34309 unique longitude values, 34309 unique latitude values


In [44]:
# X-Y plane rotation and space transformation to extract more spatial information
# 2-dimensional rotation based on below functions:
# rotated x = xcos - ysin
# rotated y = xsin + ycos
# Conver from cartesian space -> polar space

cos_30 = math.cos(math.radians(30))
sin_30 = math.sin(math.radians(30))
cos_45 = math.cos(math.radians(45))
sin_45 = math.sin(math.radians(45))
cos_60 = math.cos(math.radians(60))
sin_60 = math.sin(math.radians(60))


train["Rot30_X"] = train['X'] * cos_30 - train['Y'] * sin_30 
train["Rot30_Y"] = train['X'] * sin_30 + train['Y'] * cos_30
train["Rot45_X"] = train['X'] * cos_45 - train['Y'] * sin_45  
train["Rot45_Y"] = train['X'] * sin_45 + train['Y'] * cos_45
train["Rot60_X"] = train['X'] * cos_60 - train['Y'] * sin_60  
train["Rot60_Y"] = train['X'] * sin_60 + train['Y'] * cos_60
train["Radius"] = np.sqrt(train['X'] ** 2 + train['Y'] ** 2)
train["Angle"] = np.arctan2(train['X'], train['Y'])

test["Rot30_X"] = test['X'] * cos_30 - test['Y'] * sin_30  
test["Rot30_Y"] = test['X'] * sin_30 + test['Y'] * cos_30
test["Rot45_X"] = test['X'] * cos_45 - test['Y'] * sin_45  
test["Rot45_Y"] = test['X'] * sin_45 + test['Y'] * cos_45
test["Rot60_X"] = test['X'] * cos_60 - test['Y'] * sin_60  
test["Rot60_Y"] = test['X'] * sin_60 + test['Y'] * cos_60
test["Radius"] = np.sqrt(test['X'] ** 2 + test['Y'] ** 2)
test["Angle"] = np.arctan2(test['X'], test['Y'])

In [45]:
train.describe()

,Dates,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,BusinessHour,Quarter,Season,Weekend,Block,BlockNo,Rot30_X,Rot30_Y,Rot45_X,Rot45_Y,Rot60_X,Rot60_Y,Radius,Angle
count,878049,8.780490e+05,8.780490e+05,878049.000000,878049.000000,878049.000000,878049.000000,878049.000000,878049.000000,878049.000000,878049.000000,878049.000000,878049.000000,878049.000000,878049.000000,878049.000000,8.780490e+05,8.780490e+05,8.780490e+05,8.780490e+05,8.780490e+05,8.780490e+05,878049.000000,878049.000000
mean,2009-03-16 08:25:41.991847168,-3.892176e-13,5.497464e-13,20.155026,13.412655,15.570623,6.436509,2008.712046,2.777642,12.185458,0.518439,2.485547,2.513081,0.277339,0.702957,7.743576,-6.119538e-13,2.814893e-13,-6.639471e-13,1.135167e-13,-6.707078e-13,-6.220067e-14,1.219284,0.278898
min,2003-01-06 00:01:00,-5.291793e+00,-4.445911e+00,0.000000,0.000000,1.000000,1.000000,2003.000000,0.000000,0.000000,0.000000,1.000000,1.000000,0.000000,0.000000,0.000000,-5.504646e+00,-4.952217e+00,-5.045511e+00,-5.405149e+00,-4.242533e+00,-5.579461e+00,0.020075,-3.141532
25%,2006-01-11 03:00:00,-4.028427e-01,-6.043690e-01,0.000000,9.000000,8.000000,3.000000,2006.000000,2.000000,6.000000,0.000000,1.000000,2.000000,0.000000,0.000000,0.000000,-4.145725e-01,-5.664582e-01,-4.914589e-01,-5.402216e-01,-5.844074e-01,-4.589335e-01,0.742421,-1.065869
50%,2009-03-07 16:00:00,2.508240e-01,3.470608e-01,19.000000,14.000000,16.000000,6.000000,2009.000000,3.000000,12.000000,1.000000,2.000000,2.000000,0.000000,1.000000,4.000000,6.282368e-02,2.142762e-01,-9.705092e-02,2.239755e-01,-2.373120e-01,3.122597e-01,1.003685,0.508738
75%,2012-06-11 10:13:00,6.250366e-01,7.172741e-01,33.000000,19.000000,23.000000,9.000000,2012.000000,4.000000,19.000000,1.000000,3.000000,4.000000,1.000000,1.000000,11.000000,4.804432e-01,8.071929e-01,4.085395e-01,7.867508e-01,5.063188e-01,7.512144e-01,1.639650,1.172389
max,2015-05-13 23:53:00,3.533345e+00,3.543186e+00,59.000000,23.000000,31.000000,12.000000,2015.000000,4.000000,25.000000,1.000000,4.000000,4.000000,1.000000,1.000000,84.000000,4.848199e+00,3.779249e+00,5.027395e+00,3.661755e+00,4.863982e+00,3.294718e+00,5.603755,3.141197
std,NaN,1.000001e+00,1.000001e+00,18.594915,6.549573,8.783005,3.428972,3.631194,1.142067,7.482940,0.499660,1.117839,1.113427,0.447685,0.456956,10.233725,9.308556e-01,1.064664e+00,9.196947e-01,1.074320e+00,9.308556e-01,1.064664e+00,0.716482,1.652430


In [46]:
# run KMeans separately on both the training set and test set
data = [train, test]
num_clusters = 40
for dataset in data:
    coordinates = dataset.loc[:,['Y','X']]
    kmeans = KMeans(n_clusters=num_clusters, random_state=1).fit(coordinates)
    id_labels=kmeans.labels_
    # print(kmeans.cluster_centers_)
    dataset['Cluster'] = id_labels

In [47]:
train.sample(5)

,Dates,Category,Descript,DayOfWeek,PdDistrict,Resolution,Address,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour,Quarter,Season,Weekend,StreetType,Block,BlockNo,Rot30_X,Rot30_Y,Rot45_X,Rot45_Y,Rot60_X,Rot60_Y,Radius,Angle,Cluster
547349,2007-08-23 00:01:00,SUSPICIOUS OCC,SUSPICIOUS OCCURRENCE,Thursday,SOUTHERN,NONE,900 Block of HOWARD ST,0.642839,0.550255,1,0,23,8,2007,0,16,False,0,3,3,0,INT,1,10,0.281587,0.797954,0.065467,0.843644,-0.155115,0.831842,0.846181,0.862843,20
241322,2012-02-12 03:45:00,VANDALISM,"MALICIOUS MISCHIEF, VANDALISM",Sunday,CENTRAL,NONE,500 Block of GRANT AV,0.665377,1.037736,45,3,12,2,2012,1,2,False,0,1,1,1,AV,1,6,0.057365,1.231394,-0.263298,1.204283,-0.566017,1.095101,1.232730,0.570151,26
624857,2006-07-01 09:00:00,OTHER OFFENSES,LOST/STOLEN LICENSE PLATE,Saturday,BAYVIEW,NONE,900 Block of VERMONT ST,0.764066,-0.385945,0,9,1,7,2006,2,12,False,1,3,3,1,ST,1,10,0.854673,0.047795,0.813181,0.267372,0.716271,0.468728,0.856009,2.038531,19
254416,2011-12-04 22:00:00,LARCENY/THEFT,PETTY THEFT FROM A BUILDING,Sunday,RICHMOND,NONE,300 Block of CLEMENT ST,-1.584329,0.663500,0,22,4,12,2011,4,23,False,0,4,1,1,ST,1,4,-1.703819,-0.217556,-1.589455,-0.651124,-1.366772,-1.040319,1.717652,-1.174198,4
376250,2010-02-11 22:00:00,VEHICLE THEFT,STOLEN TRUCK,Thursday,PARK,NONE,600 Block of ASHBURY ST,-0.954658,0.103036,0,22,11,2,2010,4,2,False,0,1,1,0,ST,1,7,-0.878276,-0.388098,-0.747902,-0.602188,-0.566561,-0.775241,0.960203,-1.463283,11


# Drop Features
- We have already extracted all the necessary features from the Address attribute, so drop
- We don't need Resolution or Descript features since it is not included in the training data

In [48]:
# Drop Address feature from both train and test set
train.drop(['Address'], axis=1, inplace=True)
test.drop(['Address'], axis=1, inplace=True)

In [49]:
# We don't need Dates column anymore
train.drop(['Dates'], axis=1, inplace=True)
test.drop(['Dates'], axis=1, inplace=True)

In [50]:
# Drop Resolution column since test set does not have this column
train.drop(['Resolution'], axis=1, inplace=True)

In [51]:
# Drop Descript column since test set does not have this column
train.drop(['Descript'], axis=1, inplace=True)

In [52]:
# Let's quickly view the data
train.head()

,Category,DayOfWeek,PdDistrict,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour,Quarter,Season,Weekend,StreetType,Block,BlockNo,Rot30_X,Rot30_Y,Rot45_X,Rot45_Y,Rot60_X,Rot60_Y,Radius,Angle,Cluster
0,WARRANTS,Wednesday,NORTHERN,-0.123661,0.313043,53,23,13,5,2015,4,9,False,0,2,2,0,INT,0,0,-0.263615,0.209273,-0.308797,0.133913,-0.332934,0.049428,0.336583,-0.376214,0
1,OTHER OFFENSES,Wednesday,NORTHERN,-0.123661,0.313043,53,23,13,5,2015,4,9,False,0,2,2,0,INT,0,0,-0.263615,0.209273,-0.308797,0.133913,-0.332934,0.049428,0.336583,-0.376214,0
2,OTHER OFFENSES,Wednesday,NORTHERN,-0.063211,1.381259,33,23,13,5,2015,4,9,False,0,2,2,0,INT,0,0,-0.745372,1.164600,-1.021395,0.932001,-1.227812,0.635887,1.382705,-0.045732,23
3,LARCENY/THEFT,Wednesday,NORTHERN,-0.167305,1.400224,30,23,13,5,2015,4,9,False,0,2,2,0,INT,1,16,-0.845002,1.128977,-1.108410,0.871805,-1.296282,0.555222,1.410183,-0.118921,16
4,LARCENY/THEFT,Wednesday,PARK,-0.631649,0.186531,30,23,13,5,2015,4,9,False,0,2,2,0,ST,1,2,-0.640290,-0.154283,-0.578541,-0.314746,-0.477365,-0.453758,0.658615,-1.283649,13


# Feature Encoding
- Convert categorical data to numeric data
## Pd Districts
- convert Pd District categorical feature to numeric

In [53]:
pd_districts = {'SOUTHERN':0, 'MISSION':1, 'NORTHERN':2, 'CENTRAL':3, 'BAYVIEW':4, 'INGLESIDE':5, 
                'TENDERLOIN':6, 'TARAVAL':7, 'PARK':8, 'RICHMOND':9}

train['PdDistrict'].replace(pd_districts, inplace=True)
test['PdDistrict'].replace(pd_districts, inplace=True)

In [54]:
data = [train, test]

for dataset in data:
    year_le = LabelEncoder()
    year_le.fit(dataset['Years'].unique())
    print(list(year_le.classes_))

    dataset['Years']=year_le.transform(dataset['Years']) 

[np.int64(2003), np.int64(2004), np.int64(2005), np.int64(2006), np.int64(2007), np.int64(2008), np.int64(2009), np.int64(2010), np.int64(2011), np.int64(2012), np.int64(2013), np.int64(2014), np.int64(2015)]
[np.int64(2003), np.int64(2004), np.int64(2005), np.int64(2006), np.int64(2007), np.int64(2008), np.int64(2009), np.int64(2010), np.int64(2011), np.int64(2012), np.int64(2013), np.int64(2014), np.int64(2015)]


In [55]:
train['Years'].unique()

array([12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0])

In [56]:
# So we know the mapping (important)
dict(zip(year_le.classes_, year_le.transform(year_le.classes_)))

{np.int64(2003): np.int64(0),
 np.int64(2004): np.int64(1),
 np.int64(2005): np.int64(2),
 np.int64(2006): np.int64(3),
 np.int64(2007): np.int64(4),
 np.int64(2008): np.int64(5),
 np.int64(2009): np.int64(6),
 np.int64(2010): np.int64(7),
 np.int64(2011): np.int64(8),
 np.int64(2012): np.int64(9),
 np.int64(2013): np.int64(10),
 np.int64(2014): np.int64(11),
 np.int64(2015): np.int64(12)}

## DayOfWeek
- we are going to use sklearn's LabelEncoder to encode the categorical data to numeric
- Day of week is considered a categorical and nominal variable

In [57]:
data = [train, test]

for dataset in data:
    dow_le = LabelEncoder()
    dow_le.fit(dataset['DayOfWeek'].unique())
    print(list(dow_le.classes_))
    dataset['DayOfWeek']=dow_le.transform(dataset['DayOfWeek'])

['Friday', 'Monday', 'Saturday', 'Sunday', 'Thursday', 'Tuesday', 'Wednesday']
['Friday', 'Monday', 'Saturday', 'Sunday', 'Thursday', 'Tuesday', 'Wednesday']


In [58]:
train['DayOfWeek'].unique()

array([6, 5, 1, 3, 2, 0, 4])

In [59]:
# So we know the mapping (important)
dict(zip(dow_le.classes_, dow_le.transform(dow_le.classes_)))

{'Friday': np.int64(0),
 'Monday': np.int64(1),
 'Saturday': np.int64(2),
 'Sunday': np.int64(3),
 'Thursday': np.int64(4),
 'Tuesday': np.int64(5),
 'Wednesday': np.int64(6)}

In [60]:
train.tail(5)

,Category,DayOfWeek,PdDistrict,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour,Quarter,Season,Weekend,StreetType,Block,BlockNo,Rot30_X,Rot30_Y,Rot45_X,Rot45_Y,Rot60_X,Rot60_Y,Radius,Angle,Cluster
878044,ROBBERY,1,7,-1.434225,-2.192108,15,0,6,1,0,0,0,False,0,1,1,0,INT,0,0,-0.146021,-2.615534,0.535905,-2.564205,1.181309,-2.338129,2.619607,-2.562223,21
878045,LARCENY/THEFT,1,5,-0.972756,-1.451793,1,0,6,1,0,0,0,False,0,1,1,0,ST,1,7,-0.116535,-1.743668,0.338731,-1.714415,0.770912,-1.568328,1.747558,-2.551260,39
878046,LARCENY/THEFT,1,0,0.766143,0.547534,1,0,6,1,0,0,0,False,0,1,1,0,INT,0,0,0.389732,0.857250,0.154580,0.928910,-0.091107,0.937267,0.941684,0.950297,31
878047,VANDALISM,1,0,1.274645,0.561665,1,0,6,1,0,0,0,False,0,1,1,0,INT,0,0,0.823043,1.123739,0.504153,1.298468,0.150906,1.384708,1.392907,1.155750,15
878048,FORGERY/COUNTERFEITING,1,4,1.100874,-1.192599,1,0,6,1,0,0,0,False,0,1,1,0,AV,1,19,1.549684,-0.482384,1.621730,-0.064859,1.583258,0.357085,1.623027,2.396167,36


In [61]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 29 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Category      878049 non-null  object 
 1   DayOfWeek     878049 non-null  int64  
 2   PdDistrict    878049 non-null  int64  
 3   X             878049 non-null  float64
 4   Y             878049 non-null  float64
 5   Minutes       878049 non-null  int64  
 6   Hours         878049 non-null  int64  
 7   Days          878049 non-null  int64  
 8   Months        878049 non-null  int64  
 9   Years         878049 non-null  int64  
 10  HourZone      878049 non-null  int64  
 11  WeekOfYear    878049 non-null  int64  
 12  Holiday       878049 non-null  bool   
 13  BusinessHour  878049 non-null  uint8  
 14  Quarter       878049 non-null  int64  
 15  Season        878049 non-null  int64  
 16  Weekend       878049 non-null  uint8  
 17  StreetType    878049 non-null  object 
 18  Bloc

## Street Type
- we are going to use sklearn's LabelEncoder to encode the categorical data to numeric

In [62]:
data = [train, test]

for dataset in data:
    st_le = LabelEncoder()
    st_le.fit(dataset['StreetType'].unique())
    print(list(st_le.classes_))
    dataset['StreetType']=st_le.transform(dataset['StreetType'])

['AL', 'AV', 'BL', 'CR', 'CT', 'DR', 'EL CAMINO DEL MAR', 'HY', 'I-80', 'INT', 'LN', 'OTHER', 'PL', 'PZ', 'RD', 'RW', 'ST', 'TR', 'WAY', 'WK', 'WY']
['AL', 'AV', 'BL', 'CR', 'CT', 'DR', 'EL CAMINO DEL MAR', 'HY', 'I-80', 'INT', 'LN', 'OTHER', 'PL', 'PZ', 'RD', 'RW', 'ST', 'TR', 'WAY', 'WK', 'WY']


In [63]:
train['StreetType'].unique()

array([ 9, 16,  1, 20,  4,  5, 17,  7, 14, 13,  2, 12, 10, 18,  3,  0, 11,
       19, 15,  6,  8])

### Holiday
Encode the binary feature

In [64]:
test.sample(5)

,Id,DayOfWeek,PdDistrict,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour,Quarter,Season,Weekend,StreetType,Block,BlockNo,Rot30_X,Rot30_Y,Rot45_X,Rot45_Y,Rot60_X,Rot60_Y,Radius,Angle,Cluster
792985,792985,2,0,1.352723,0.685614,0,17,13,3,1,3,5,False,1,1,2,1,16,1,7,0.828685,1.270120,0.471717,1.441322,0.082602,1.514299,1.516550,1.101692,15
105048,105048,2,5,-0.785455,-2.101491,30,17,21,12,10,3,25,False,1,4,1,1,1,1,1,0.370521,-2.212672,0.930578,-2.041379,1.427217,-1.730969,2.243480,-2.783909,30
11726,11726,3,3,0.492360,1.613490,0,15,15,3,12,3,5,False,1,1,2,1,9,0,0,-0.380349,1.643503,-0.792759,1.489060,-1.151143,1.233141,1.686940,0.296177,7
390400,390400,2,7,-1.911206,-1.138685,0,1,21,11,6,1,23,False,0,4,4,1,9,0,0,-1.085810,-1.941733,-0.546255,-2.156599,0.030527,-2.224495,2.224705,-2.108118,18
777615,777615,1,7,-2.221388,-2.149561,30,13,31,5,1,3,11,True,1,2,2,0,5,1,1,-0.848998,-2.972269,-0.050789,-3.090728,0.750881,-2.998559,3.091145,-2.339763,27


In [65]:
# Encode to 0 and 1

train['Holiday'].replace(False, 0, inplace=True)
train['Holiday'].replace(True, 1, inplace=True)
test['Holiday'].replace(False, 0, inplace=True)
test['Holiday'].replace(True, 1, inplace=True)

train['Holiday'] = train['Holiday'].astype('uint8')
train['Holiday'] = train['Holiday'].astype('uint8')
train[train['Holiday'] == 1].head()

,Category,DayOfWeek,PdDistrict,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour,Quarter,Season,Weekend,StreetType,Block,BlockNo,Rot30_X,Rot30_Y,Rot45_X,Rot45_Y,Rot60_X,Rot60_Y,Radius,Angle,Cluster
18357,DRUG/NARCOTIC,1,0,0.765573,0.347061,48,23,16,2,12,4,3,1,0,1,1,0,16,1,9,0.489475,0.683350,0.295933,0.786751,0.082223,0.836536,0.840567,1.145173,31
18358,OTHER OFFENSES,1,0,0.765573,0.347061,48,23,16,2,12,4,3,1,0,1,1,0,16,1,9,0.489475,0.683350,0.295933,0.786751,0.082223,0.836536,0.840567,1.145173,31
18359,NON-CRIMINAL,1,5,0.436357,-2.363600,30,23,16,2,12,4,3,1,0,1,1,0,16,1,3,1.559696,-1.828759,1.979868,-1.362766,2.265116,-0.803904,2.403541,2.959033,30
18360,VANDALISM,1,3,0.281625,1.685316,30,23,16,2,12,4,3,1,0,1,1,0,16,1,28,-0.598764,1.600339,-0.992559,1.390837,-1.318714,1.086552,1.708684,0.165575,23
18361,LARCENY/THEFT,1,2,-0.250930,0.118574,30,23,16,2,12,4,3,1,0,1,1,0,16,1,1,-0.276598,-0.022777,-0.261279,-0.093590,-0.228153,-0.158025,0.277535,-1.129359,13


In [66]:
test[test['Holiday'] == 1].head()


,Id,DayOfWeek,PdDistrict,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour,Quarter,Season,Weekend,StreetType,Block,BlockNo,Rot30_X,Rot30_Y,Rot45_X,Rot45_Y,Rot60_X,Rot60_Y,Radius,Angle,Cluster
27957,27957,4,3,0.499739,1.575083,55,23,1,1,12,4,0,1,0,1,1,0,9,0,0,-0.354755,1.613931,-0.760384,1.467121,-1.114193,1.220328,1.652461,0.307231,7
27958,27958,4,6,0.192209,0.750352,43,23,1,1,12,4,0,1,0,1,1,0,9,0,0,-0.208718,0.745928,-0.394667,0.666491,-0.553719,0.541633,0.774578,0.250766,4
27959,27959,4,1,-0.121703,-0.265287,30,23,1,1,12,4,0,1,0,1,1,0,16,1,6,0.027245,-0.290597,0.101529,-0.273644,0.168894,-0.238042,0.291872,-2.711477,8
27960,27960,4,7,-2.341104,-1.309983,15,23,1,1,12,4,0,1,0,1,1,0,9,1,1,-1.372464,-2.305031,-0.729113,-2.581709,-0.036074,-2.682447,2.682690,-2.080948,18
27961,27961,4,7,-2.341104,-1.309983,15,23,1,1,12,4,0,1,0,1,1,0,9,1,1,-1.372464,-2.305031,-0.729113,-2.581709,-0.036074,-2.682447,2.682690,-2.080948,18


In [67]:
data = [train]

for dataset in data:
    cat_le = LabelEncoder()
    cat_le.fit(dataset['Category'].unique())
    print(list(cat_le.classes_))
    dataset['Category']=cat_le.transform(dataset['Category'])

['ARSON', 'ASSAULT', 'BAD CHECKS', 'BRIBERY', 'BURGLARY', 'DISORDERLY CONDUCT', 'DRIVING UNDER THE INFLUENCE', 'DRUG/NARCOTIC', 'DRUNKENNESS', 'EMBEZZLEMENT', 'EXTORTION', 'FAMILY OFFENSES', 'FORGERY/COUNTERFEITING', 'FRAUD', 'GAMBLING', 'KIDNAPPING', 'LARCENY/THEFT', 'LIQUOR LAWS', 'LOITERING', 'MISSING PERSON', 'NON-CRIMINAL', 'OTHER OFFENSES', 'PORNOGRAPHY/OBSCENE MAT', 'PROSTITUTION', 'RECOVERED VEHICLE', 'ROBBERY', 'RUNAWAY', 'SECONDARY CODES', 'SEX OFFENSES FORCIBLE', 'SEX OFFENSES NON FORCIBLE', 'STOLEN PROPERTY', 'SUICIDE', 'SUSPICIOUS OCC', 'TREA', 'TRESPASS', 'VANDALISM', 'VEHICLE THEFT', 'WARRANTS', 'WEAPON LAWS']


In [68]:
len(train['Category'].unique())

39

In [69]:
# So we know the mapping (important)
dict(zip(cat_le.classes_, cat_le.transform(cat_le.classes_)))

{'ARSON': np.int64(0),
 'ASSAULT': np.int64(1),
 'BAD CHECKS': np.int64(2),
 'BRIBERY': np.int64(3),
 'BURGLARY': np.int64(4),
 'DISORDERLY CONDUCT': np.int64(5),
 'DRIVING UNDER THE INFLUENCE': np.int64(6),
 'DRUG/NARCOTIC': np.int64(7),
 'DRUNKENNESS': np.int64(8),
 'EMBEZZLEMENT': np.int64(9),
 'EXTORTION': np.int64(10),
 'FAMILY OFFENSES': np.int64(11),
 'FORGERY/COUNTERFEITING': np.int64(12),
 'FRAUD': np.int64(13),
 'GAMBLING': np.int64(14),
 'KIDNAPPING': np.int64(15),
 'LARCENY/THEFT': np.int64(16),
 'LIQUOR LAWS': np.int64(17),
 'LOITERING': np.int64(18),
 'MISSING PERSON': np.int64(19),
 'NON-CRIMINAL': np.int64(20),
 'OTHER OFFENSES': np.int64(21),
 'PORNOGRAPHY/OBSCENE MAT': np.int64(22),
 'PROSTITUTION': np.int64(23),
 'RECOVERED VEHICLE': np.int64(24),
 'ROBBERY': np.int64(25),
 'RUNAWAY': np.int64(26),
 'SECONDARY CODES': np.int64(27),
 'SEX OFFENSES FORCIBLE': np.int64(28),
 'SEX OFFENSES NON FORCIBLE': np.int64(29),
 'STOLEN PROPERTY': np.int64(30),
 'SUICIDE': np.int6

In [70]:
train.head(5)

,Category,DayOfWeek,PdDistrict,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour,Quarter,Season,Weekend,StreetType,Block,BlockNo,Rot30_X,Rot30_Y,Rot45_X,Rot45_Y,Rot60_X,Rot60_Y,Radius,Angle,Cluster
0,37,6,2,-0.123661,0.313043,53,23,13,5,12,4,9,0,0,2,2,0,9,0,0,-0.263615,0.209273,-0.308797,0.133913,-0.332934,0.049428,0.336583,-0.376214,0
1,21,6,2,-0.123661,0.313043,53,23,13,5,12,4,9,0,0,2,2,0,9,0,0,-0.263615,0.209273,-0.308797,0.133913,-0.332934,0.049428,0.336583,-0.376214,0
2,21,6,2,-0.063211,1.381259,33,23,13,5,12,4,9,0,0,2,2,0,9,0,0,-0.745372,1.164600,-1.021395,0.932001,-1.227812,0.635887,1.382705,-0.045732,23
3,16,6,2,-0.167305,1.400224,30,23,13,5,12,4,9,0,0,2,2,0,9,1,16,-0.845002,1.128977,-1.108410,0.871805,-1.296282,0.555222,1.410183,-0.118921,16
4,16,6,8,-0.631649,0.186531,30,23,13,5,12,4,9,0,0,2,2,0,16,1,2,-0.640290,-0.154283,-0.578541,-0.314746,-0.477365,-0.453758,0.658615,-1.283649,13


In [71]:
train.sample(5)

,Category,DayOfWeek,PdDistrict,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour,Quarter,Season,Weekend,StreetType,Block,BlockNo,Rot30_X,Rot30_Y,Rot45_X,Rot45_Y,Rot60_X,Rot60_Y,Radius,Angle,Cluster
148882,1,1,3,0.394746,1.598544,10,19,27,5,10,4,10,1,0,2,2,0,16,1,4,-0.457412,1.581753,-0.851214,1.409469,-1.187007,1.141132,1.646562,0.242097,23
578610,18,4,2,-0.247278,0.277195,18,8,8,3,4,2,4,0,1,1,2,0,9,1,4,-0.352746,0.116419,-0.370858,0.021155,-0.363697,-0.075551,0.371461,-0.728417,33
735994,16,2,0,0.765573,0.347061,0,23,27,11,1,4,23,0,0,4,4,1,16,1,9,0.489475,0.683350,0.295933,0.786751,0.082223,0.836536,0.840567,1.145173,31
295093,4,3,4,0.855605,-2.257767,15,13,24,4,8,3,7,0,1,2,2,1,2,1,22,1.869859,-1.527482,2.201487,-0.991479,2.383086,-0.387908,2.414451,2.779354,17
467021,21,6,5,-0.691587,-2.110024,30,8,15,10,5,2,20,0,1,4,4,0,9,0,0,0.456080,-2.173128,1.002986,-1.981038,1.481541,-1.653944,2.220472,-2.824864,8


In [72]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 29 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Category      878049 non-null  int64  
 1   DayOfWeek     878049 non-null  int64  
 2   PdDistrict    878049 non-null  int64  
 3   X             878049 non-null  float64
 4   Y             878049 non-null  float64
 5   Minutes       878049 non-null  int64  
 6   Hours         878049 non-null  int64  
 7   Days          878049 non-null  int64  
 8   Months        878049 non-null  int64  
 9   Years         878049 non-null  int64  
 10  HourZone      878049 non-null  int64  
 11  WeekOfYear    878049 non-null  int64  
 12  Holiday       878049 non-null  uint8  
 13  BusinessHour  878049 non-null  uint8  
 14  Quarter       878049 non-null  int64  
 15  Season        878049 non-null  int64  
 16  Weekend       878049 non-null  uint8  
 17  StreetType    878049 non-null  int64  
 18  Bloc

### View Information about Data
One last check before training

In [73]:
# # Convert all to 32 bit integers so less memory and will train faster (no loss in data since our integers dont reach)
columns_to_convert = ['DayOfWeek', 'PdDistrict', 'Minute', 'Hour', 'Day', 'Month', 'Year', 
                      'Hour_Zone', 'WeekOfYear', 'Season', 'StreetType', 'BlockNo', 'Cluster']
# train[columns_to_convert] = train[columns_to_convert].astype('int16')
# test[columns_to_convert] = test[columns_to_convert].astype('int16')

train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 29 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Category      878049 non-null  int64  
 1   DayOfWeek     878049 non-null  int64  
 2   PdDistrict    878049 non-null  int64  
 3   X             878049 non-null  float64
 4   Y             878049 non-null  float64
 5   Minutes       878049 non-null  int64  
 6   Hours         878049 non-null  int64  
 7   Days          878049 non-null  int64  
 8   Months        878049 non-null  int64  
 9   Years         878049 non-null  int64  
 10  HourZone      878049 non-null  int64  
 11  WeekOfYear    878049 non-null  int64  
 12  Holiday       878049 non-null  uint8  
 13  BusinessHour  878049 non-null  uint8  
 14  Quarter       878049 non-null  int64  
 15  Season        878049 non-null  int64  
 16  Weekend       878049 non-null  uint8  
 17  StreetType    878049 non-null  int64  
 18  Bloc

# Building Machine Learning Models
- Baseline Models
    - Let's train a couple models on a stratified sample of the training data
    - Evaluate on a hold out set to get baseline results for each model to determine what model to use
    - Models:
        - Stochastic Gradient Descent (with elastic net regularization)
        - Gaussian Naive Bayes
        - K Nearest Neighbors
        - Logistic Regression (with L1 regularization)
        - Random Forest
        - XGBoost
    - Almost all the default scikit-learn ML algorithm hyperparameters exhibit bad performance
Researched online & read literature to determine some more ideal default hyperparameters
Reference
- Couple things to note:
    - Decision tree models including Ensemble methods (Random Forest & XGBoost) can handle categorical variables without one-hot encoding them.
    - Linear models (SGD & Logistic Regression) cannot handle categorical features & need features to be OHE before training
    - Always OneHotEncode before you split data up to training/dev/test so that all features & classes will be represented

In [74]:
# Set training data (drop labels) and training labels
X_train = train.drop("Category", axis=1).copy()
Y_train = train["Category"].copy()

# Set testing data (drop Id)
X_test = test.drop("Id", axis=1).copy()

In [75]:
def one_hot_encode(train_data):
    '''One Hot Encode the categorical features'''
    encoded_train_data = train_data

    encoded_train_data = pd.concat([encoded_train_data, 
                                    pd.get_dummies(pd.Series(encoded_train_data['PdDistrict']), prefix='PdDistrict')], axis=1)
    encoded_train_data = pd.concat([encoded_train_data, 
                                    pd.get_dummies(pd.Series(encoded_train_data['DayOfWeek']), prefix='DayOfWeek')], axis=1)
    encoded_train_data = pd.concat([encoded_train_data, 
                                    pd.get_dummies(pd.Series(encoded_train_data['StreetType']), prefix='StreetType')], axis=1)
    encoded_train_data = pd.concat([encoded_train_data, 
                                    pd.get_dummies(pd.Series(encoded_train_data['Season']), prefix='Season')], axis=1)
    encoded_train_data = pd.concat([encoded_train_data, 
                                    pd.get_dummies(pd.Series(encoded_train_data['HourZone']), prefix='HourZone')], axis=1)
    encoded_train_data = pd.concat([encoded_train_data, 
                                    pd.get_dummies(pd.Series(encoded_train_data['Cluster']), prefix='Cluster')], axis=1)
    encoded_train_data = encoded_train_data.drop(['Cluster','StreetType', 'Season', 'HourZone', 'DayOfWeek', 'PdDistrict'], axis=1)

    return encoded_train_data

In [76]:
X_train.head()

,DayOfWeek,PdDistrict,X,Y,Minutes,Hours,Days,Months,Years,HourZone,WeekOfYear,Holiday,BusinessHour,Quarter,Season,Weekend,StreetType,Block,BlockNo,Rot30_X,Rot30_Y,Rot45_X,Rot45_Y,Rot60_X,Rot60_Y,Radius,Angle,Cluster
0,6,2,-0.123661,0.313043,53,23,13,5,12,4,9,0,0,2,2,0,9,0,0,-0.263615,0.209273,-0.308797,0.133913,-0.332934,0.049428,0.336583,-0.376214,0
1,6,2,-0.123661,0.313043,53,23,13,5,12,4,9,0,0,2,2,0,9,0,0,-0.263615,0.209273,-0.308797,0.133913,-0.332934,0.049428,0.336583,-0.376214,0
2,6,2,-0.063211,1.381259,33,23,13,5,12,4,9,0,0,2,2,0,9,0,0,-0.745372,1.164600,-1.021395,0.932001,-1.227812,0.635887,1.382705,-0.045732,23
3,6,2,-0.167305,1.400224,30,23,13,5,12,4,9,0,0,2,2,0,9,1,16,-0.845002,1.128977,-1.108410,0.871805,-1.296282,0.555222,1.410183,-0.118921,16
4,6,8,-0.631649,0.186531,30,23,13,5,12,4,9,0,0,2,2,0,16,1,2,-0.640290,-0.154283,-0.578541,-0.314746,-0.477365,-0.453758,0.658615,-1.283649,13


In [77]:
X_encoded_train = one_hot_encode(X_train)
X_encoded_train.sample(5)

,X,Y,Minutes,Hours,Days,Months,Years,WeekOfYear,Holiday,BusinessHour,Quarter,Weekend,Block,BlockNo,Rot30_X,Rot30_Y,Rot45_X,Rot45_Y,Rot60_X,Rot60_Y,Radius,Angle,PdDistrict_0,PdDistrict_1,PdDistrict_2,PdDistrict_3,PdDistrict_4,PdDistrict_5,PdDistrict_6,PdDistrict_7,PdDistrict_8,PdDistrict_9,DayOfWeek_0,DayOfWeek_1,DayOfWeek_2,DayOfWeek_3,DayOfWeek_4,DayOfWeek_5,DayOfWeek_6,StreetType_0,StreetType_1,StreetType_2,StreetType_3,StreetType_4,StreetType_5,StreetType_6,StreetType_7,StreetType_8,StreetType_9,StreetType_10,StreetType_11,StreetType_12,StreetType_13,StreetType_14,StreetType_15,StreetType_16,StreetType_17,StreetType_18,StreetType_19,StreetType_20,Season_1,Season_2,Season_3,Season_4,HourZone_0,HourZone_1,HourZone_2,HourZone_3,HourZone_4,Cluster_0,Cluster_1,Cluster_2,Cluster_3,Cluster_4,Cluster_5,Cluster_6,Cluster_7,Cluster_8,Cluster_9,Cluster_10,Cluster_11,Cluster_12,Cluster_13,Cluster_14,Cluster_15,Cluster_16,Cluster_17,Cluster_18,Cluster_19,Cluster_20,Cluster_21,Cluster_22,Cluster_23,Cluster_24,Cluster_25,Cluster_26,Cluster_27,Cluster_28,Cluster_29,Cluster_30,Cluster_31,Cluster_32,Cluster_33,Cluster_34,Cluster_35,Cluster_36,Cluster_37,Cluster_38,Cluster_39
367413,0.606200,0.847541,20,15,5,4,7,6,0,1,2,0,1,3,0.101215,1.037092,-0.170653,1.027950,-0.430892,0.948755,1.042019,0.620885,False,False,False,True,False,False,False,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,True,False,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False
826304,-0.851150,0.095217,30,17,7,9,0,17,0,1,3,1,1,13,-0.784726,-0.343115,-0.669183,-0.534526,-0.508035,-0.689509,0.856460,-1.459391,False,False,False,False,False,False,False,False,True,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
689901,-1.364060,-0.087264,0,14,31,7,2,14,0,1,3,1,1,5,-1.137679,-0.757603,-0.902831,-1.026241,-0.606457,-1.224943,1.366849,-1.634683,False,False,False,False,False,False,False,False,True,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False,True,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
334604,-0.891197,-2.276585,54,16,24,9,7,18,0,1,3,0,1,56,0.366493,-2.417178,0.979617,-2.239959,1.525982,-1.910091,2.444804,-2.768468,False,False,False,False,False,True,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False,False,True,False,False,False,True,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
692965,0.513874,-0.339981,0,8,16,7,2,13,0,1,3,1,1,23,0.615018,-0.037495,0.603767,0.122961,0.551369,0.275037,0.616160,2.155286,False,True,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,Fa

In [78]:
# Use these for ML algorithms that can't handle categorical data (Logistic Regression, Linear Models)
mini_encoded_train_data, mini_encoded_dev_data, mini_train_labels, mini_dev_labels = train_test_split(X_encoded_train, 
                                                                                      Y_train,
                                                                                      stratify=Y_train,
                                                                                      test_size=0.5,
                                                                                      random_state=1)

In [79]:
# Use these for ML algorithms that can handle categorical data without OHE
mini_train_data, mini_dev_data, mini_train_labels, mini_dev_labels = train_test_split(X_train, 
                                                                                      Y_train,
                                                                                      stratify=Y_train,
                                                                                      test_size=0.5,
                                                                                      random_state=1)

In [80]:
# K Neighbors
knn = KNeighborsClassifier()
knn.fit(mini_train_data, mini_train_labels)
pred_probs = knn.predict_proba(mini_dev_data)
knn_loss = log_loss(mini_dev_labels, pred_probs)


print('KNN Validation Log Loss: ', knn_loss)

KNN Validation Log Loss:  18.684293090053128


In [81]:
# Naive Bayes
gaussian = GaussianNB()
gaussian.fit(mini_train_data, mini_train_labels)
pred_probs = gaussian.predict_proba(mini_dev_data)
nb_loss = log_loss(mini_dev_labels, pred_probs)


print('Gaussian Naive Bayes Validation Log Loss: ', nb_loss)

Gaussian Naive Bayes Validation Log Loss:  4.137054863569331


In [82]:
# stochastic gradient descent (SGD) learning
sgd = linear_model.SGDClassifier(penalty='elasticnet', loss='log', 
                                  tol=0.0001, max_iter=1000, n_jobs=3, random_state=1)
sgd.fit(mini_encoded_train_data, mini_train_labels)
pred_probs = sgd.predict_proba(mini_encoded_dev_data)
# sgd.fit(one_hot_encode(mini_train_data), mini_train_labels)
# sgd = gaussian.predict_proba(one_hot_encode(mini_dev_data))
sgd_loss = log_loss(mini_dev_labels, pred_probs)

print('Linear Model SGD Validation Log Loss: ', sgd_loss)

InvalidParameterError: The 'loss' parameter of SGDClassifier must be a str among {'epsilon_insensitive', 'perceptron', 'log_loss', 'huber', 'squared_hinge', 'modified_huber', 'squared_epsilon_insensitive', 'squared_error', 'hinge'}. Got 'log' instead.

In [ ]:
# Logistic Regression
logreg = LogisticRegression(penalty='l1', C=1.5, solver='saga', multi_class='multinomial', 
                            tol=0.0001, max_iter=1000, verbose=3, n_jobs=3, random_state=1)

logreg.fit(mini_encoded_train_data, mini_train_labels)
pred_probs = logreg.predict_proba(mini_encoded_dev_data)

logreg_loss = log_loss(mini_dev_labels, pred_probs)


print('Logistic Regression Validation Log Loss: ', logreg_loss)

[Parallel(n_jobs=3)]: Using backend ThreadingBackend with 3 concurrent workers.
